### PR TITLE
Fixed backpropagation for models using parameters.grad

### PR DESCRIPTION
**Pull request by DanyEle**
Fix for issue #2122 reported while trying to implement a Recurrent Neural Network in a federated way. 

Fixed the loss.backward() function not generating gradient updates accumulated in the model parameters.

If trying to access the .grad.data attribute of a model's parameters, not all the gradient parameters were actually set, leading to the model weights not being updated. The de-commented lines actually allow the gradients to be stored in the model parameters and the weights to be updated, hence the loss function to converge.

### CODE REVIEW

**Comment by robert-wagner** [comment]
Hey, This is a super high value pull request. It looks like flake8 is not happy with indentation which is causing travis to fail. That being said could you add a test to make sure this is not broken again in the future

**Comment by DanyEle** [comment]
Sorry for the many commits, yesterday night I messed up badly with Git. Going to try and see if I can get a test up for this fix.

File: test/torch/test_hook.py [hunk 1]
```diff
@@ -220,3 +220,69 @@ def test_hook_args_and_cmd_signature_malleability():
 
     r3 = a + b
     assert (r3 == syft.LoggingTensor().on(torch.tensor([2.0, 4]))).all()
+
+
+def test_RNN_grad_set_backpropagation():
+    """Perform backpropagation at a remote worker and check if the gradient updates
+    and properly computed within the model"""
+
+    import syft as sy
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: You don't need this import

File: test/torch/test_hook.py [hunk 2]
```diff
@@ -220,3 +220,69 @@ def test_hook_args_and_cmd_signature_malleability():
...
+
+    hook = sy.TorchHook(torch)
+    alice = sy.VirtualWorker(hook, id="alice")
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: You can get Alice by including workers as a parameter and writing Alice = workers["Alice"]. The tests for most tensors have good examples for how to do this

File: test/torch/test_hook.py [hunk 3]
```diff
@@ -220,3 +220,69 @@ def test_hook_args_and_cmd_signature_malleability():
...
+
+    # now let's get the model and check if its parameters are indeed there
+    model_got = alice_model.get()
+
+    learning_rate = 0.005
+
+    # If the gradients are there, then the backpropagation did indeed complete successfully
+    for param in model_got.parameters():
+        assert param.grad.data is not None
+        param.data.add_(-learning_rate, param.grad.data)
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: What does this line do?

Comment (DanyEle) [reply test/torch/test_hook.py]: This is the most important line, in which the` param.grad.data` was previously yielding a `None`. It basically updates the model's parameters (e.g.: layers' weights) based on the gradients computed following the backpropagation phase.

File: test/torch/test_hook.py (hunk 2 above)
Comment (DanyEle) [reply test/torch/test_hook.py]: Done!

File: test/torch/test_hook.py (hunk 1 above)
Comment (DanyEle) [reply test/torch/test_hook.py]: Removed import

File: test/torch/test_hook.py (hunk 3 above)
Comment (robert-wagner) [reply test/torch/test_hook.py]: Why is this called after the assertion though

Comment (DanyEle) [reply test/torch/test_hook.py]: Basically with this "assert2, we would like to make sure that every single gradient among the model's parameters is indeed there.

 By executing `param.data.add_(-learning_rate, param.grad.data)` without an an assert before-hand, we may already have an exception in case a grad.data is None, meaning that the code is broken again and we had no way to check the effective presence/absence of the gradient beforehand
